src/tv2-common/onTimelineGenerate.ts

```diff
@@ -295,53 +320,72 @@ export function createSisyfosPersistedLevelsTimelineObject(
 	}
 }
 
-function findLayersToPersist(
-	pieces: Array<IBlueprintResolvedPieceInstance<PieceMetaData>>,
-	sisyfosLayersThatWantsToBePersisted: string[]
+function findLayersToPersistOnPartEnd(
+	pieceInstances: Array<IBlueprintResolvedPieceInstance<PieceMetaData>>,
+	layersWantingToPersistFromPreviousPart: string[] = []
 ): string[] {
-	const sortedPieces = pieces
-		.filter((piece) => piece.piece.metaData?.sisyfosPersistMetaData)
-		.sort((a, b) => b.resolvedStart - a.resolvedStart)
+	const latestPieceInstance = findLastPlayingPieceInstance(
+		pieceInstances,
+		(pieceInstance) => !!pieceInstance.piece.metaData?.sisyfosPersistMetaData
+	)
+	const latestPieceMetaData = latestPieceInstance?.piece.metaData?.sisyfosPersistMetaData
 
-	if (sortedPieces.length === 0) {
+	if (!latestPieceMetaData?.wantsToPersistAudio) {
 		return []
 	}
 
-	const firstPieceMetaData = sortedPieces[0].piece.metaData!
-	if (!firstPieceMetaData.sisyfosPersistMetaData!.acceptPersistAudio) {
-		return firstPieceMetaData.sisyfosPersistMetaData!.wantsToPersistAudio
-			? firstPieceMetaData.sisyfosPersistMetaData!.sisyfosLayers
-			: []
+	if (!latestPieceMetaData.acceptsPersistedAudio) {
+		return latestPieceMetaData.wantsToPersistAudio ? latestPieceMetaData.sisyfosLayers : []
 	}
 
 	const layersToPersist: string[] = []
-	for (let i = 0; i < sortedPieces.length; i++) {
-		const pieceMetaData = sortedPieces[i].piece.metaData!
-		const sisyfosPersistMetaData: SisyfosPersistMetaData = pieceMetaData.sisyfosPersistMetaData!
-		if (sisyfosPersistMetaData.wantsToPersistAudio) {
-			layersToPersist.push(...sisyfosPersistMetaData.sisyfosLayers)
-		}
+	if (!latestPieceMetaData?.isModifiedOrInsertedByAction) {
+		layersToPersist.push(...layersWantingToPersistFromPreviousPart)
+	} else if (latestPieceMetaData.previousSisyfosLayers) {
+		layersToPersist.push(...latestPieceMetaData.previousSisyfosLayers)
+	}
+	layersToPersist.push(...latestPieceMetaData.sisyfosLayers)
 
-		if (doesMetaDataNotAcceptPersistAudioDeep(sisyfosPersistMetaData)) {
-			break
-		}
+	return Array.from(new Set(layersToPersist))
+}
 
-		if (i === sortedPieces.length - 1) {
-			layersToPersist.push(...sisyfosLayersThatWantsToBePersisted)
-		}
+function findPersistedLayers(
```

Comment:
This method and `findLayersToPersistOnPartEnd` are really similar. Would guess around 80%. Is it possible to extract some of the duplication away?